documentation/design/rar-core-scenarios.md

```diff
@@ -0,0 +1,193 @@
+
+The pre-cache functionality is generic and available to any SDK vendor. The .NET SDK currently builds and distributes a file named `SDKPrecomputedAssemblyReferences.cache`
+but it is not passed to RAR by default. Only a couple of projects in the dotnet organization are explicitly opted into consuming the pre-cache at the moment.
+
+The downside of the current pre-cache design is that the full pre-cache ends up being written to each per project cache file upon completing the first RAR invocation.
+For the .NET SDK the pre-cache contains more than 3000 assemblies. All of them stay in memory in the per process cache and all of them become part of the
+per-project cache file, meaning that they will be read back from disk on each subsequent hot invocation. Not only does it hurt build performance, but it is also
+wasteful to duplicate >2 MB worth of serialized assembly information in each project's intermediate directory.
+
+## Proposed design
+
+Completely rewriting RAR doesn't appear to be worthwhile. The requirements described above are for a typical build, not necessarily for all builds. RAR is highly
+configurable and customizable, thus the bar for backward compatibility is very high. There are definitely opportunities for micro-optimizations without any functional
+effect. Be it eliminating allocations, simplifying tight loops, reordering cases in hot switches, ..., there is a lot of low-hanging fruit. This by itself won't help
+address the elephant in the room: the file I/O resulting from scanning of assemblies, checking their timestamps, and reading/writing on-disk caches.
+
+For regular project references the system works as about as efficient as possible.
+- In a cold scenario, where there is no state in memory or on disk, the referenced assembly file has to be scanned for its name and dependencies.
+- In a warm scenario, where there is no state in memory but a disk cache exists, the assembly name and dependencies are read from the cache, together with the
+corresponding timestamp which is compared to the current timestamp of the assembly file. If they match the cached data is used.
+- In a hot scenario, where there is state in memory, the only I/O on the happy path is the timestamp check to verify that the file hasn't changed since last time.
+
+There is a chance that the timestamp check can be replaced with something faster, although historically we haven't been able to come up with anything solid.
+File watchers, for example, while tempting to use because the validity check in the happy case would cost literally nothing, suffer from an inherent race
+condition. When a watched file is modified, the file watcher routine is not guaranteed to run by the time we need to reliably know whether the file is unchanged.
+The exact time the routine is executed depends on the latency of the asynchronous OS callback, on thread pool availability, CPU scheduling, and more.
```

Comment:
Btw. this is interesting line of thoughts as well. Having a benefit of not knowing the RAR internals and seeing things oversimplified - I'm thinking that if we'd have a long running process (e.g. MSBuild server), we can collect info via FileWatchers, prior the start of the build we'd perform a dummy 'marker' change - so that we can see whether this was received and hence we can trust the state prior the build (otherwise we'd need to fallback to traditional way) and then during the build we'd need to diligently explicitly update the cache (detours tracker might be a strategy here) when changing file system state (e.g. writing build outputs) - so that we can trust the cache state.

It surely get's quite complicated, but it then solves the hard choices of honoring/ignoring changes in low-freqency changing locations (sdk ref assemblies, nuget cache). 

But I agree that this definitely isn't a low hanging fruit and it's definitely worth to explore lower cost improvements first.

Reply:
Thank you, yes, I forgot about the trick with a dummy marker change. If watcher callbacks are guaranteed to be delivered in order, it may be worth experimenting with.

In general I believe RAR and MSBuild as a whole is in a state where it gets more value from simplifying the logic, rather than adding more complexity.

Reply:
That sounds very reasonable.
I'm all for simplifications